lib/child_process.js

```diff
@@ -245,39 +246,47 @@ exports.execFile = function(file /*, args, options, callback*/) {
   }
 
   if (child.stdout) {
-    if (encoding)
-      child.stdout.setEncoding(encoding);
+    const stdoutState = child.stdout._readableState;
 
     child.stdout.addListener('data', function(chunk) {
-      stdoutLen += chunk.length;
+      // If `child.stdout.setEncoding('utf8') happened in userland, convert
+      // string to Buffer and save encoding for later.
+      if (stdoutState.decoder) {
+        stdoutEncoding = stdoutState.decoder.encoding;
+        chunk = Buffer.from(chunk, stdoutEncoding);
+      }
+
+      stdoutLen += chunk.byteLength;
 
       if (stdoutLen > options.maxBuffer) {
         ex = new Error('stdout maxBuffer exceeded');
+        stdoutLen -= chunk.byteLength;
```

Comment:
@mscdex It's needed so that it has the correct value when passed to `Buffer.concat()`. The alternative is to let the buffer exceed `options.maxBuffer`. However, that does not match the behavior of the current implementation whereas this does.